Change charms to use juju-http{s}-proxy model settings

VariableDeclared opened the pull request:
Changed charm to use juju-http{s}-proxy-settings when these keys are defined.

[LP#1831712](https://bugs.launchpad.net/charm-kubernetes-worker/+bug/1831712)

VariableDeclared committed:
Added juju-http{s}-proxy environment config

Added code to pull down https_proxy config and override
any existing keys if they exist.

VariableDeclared committed:
Documentation

VariableDeclared committed:
Add missing paranthesis

Added missing parenthesis which cause a crash at run-time.

VariableDeclared commented:
~~Pending Jenkins Tests
[#8](https://github.com/charmed-kubernetes/charm-containerd/pull/8)
[#138](https://github.com/juju-solutions/layer-docker/pull/138)~~

VariableDeclared committed:
Added proxy key logic to override defaults

Added proxy key configuration to override default
proxy options.

LP#:https://bugs.launchpad.net/charm-kubernetes-master/+bug/1831712

VariableDeclared commented:
Added Jenkins Tests; [#447](https://github.com/charmed-kubernetes/jenkins/pull/447/files)

VariableDeclared committed:
Testing docs

Added docs on how to update integration tests.

VariableDeclared committed:
Remove testing changes

Removed comment block and change to exception
statement.

VariableDeclared committed:
Refactored the configuration update

There is an issue where calling
config().update({val}) causes docker-relation-departed
to fail. Initial look into this does not reveal much.

But changing code to val = config() then updating does not
cause the issue.

Needs investigation, will write up notes.

Cynerva commented on reactive/docker.py:
```diff
@@ -919,6 +921,18 @@ def recycle_daemon():
+    environment_config = hookenv.env_proxy_settings()
+    configuration = config()
+    if environment_config is not None:
+        configuration.update(environment_config)
```
The [Config object](https://github.com/juju/charm-helpers/blob/b31408ece9f39b0bc527d785acd9ab492c14ff39/charmhelpers/core/hookenv.py#L278) returned by `config()` should be considered read-only. If the charm writes to it like this, then future hooks will get incorrect results from calls to `config.previous()`, and that will cause `config.changed.http_proxy`, `config.changed.https_proxy`, and `config.changed.no_proxy` to be set on every hook execution. That will ultimately lead to `recycle_daemon` being called on every hook.

Cynerva commented:
Does this support CIDR notation in juju-no-proxy? From the [Juju docs](https://docs.jujucharms.com/offline-mode-strategies):

> One big benefit of using these finely-scoped settings is that juju-no-proxy can contain subnets (in CIDR notation) whereas its legacy counterpart cannot.

Docker most likely does not understand CIDR notation in the NO_PROXY env var (most apps don't), meaning we'd have to expand it into individual IPs somewhere along the way. I'm looking at the definition of [env_proxy_settings](https://github.com/juju/charm-helpers/blob/b31408ece9f39b0bc527d785acd9ab492c14ff39/charmhelpers/core/hookenv.py#L1424) and it doesn't look like it expands them, so that leaves it as the responsibility of the caller to do so.

VariableDeclared committed:
Refactored method to container-runtime-common

Moved juju configuration check into the container
runtime common layer.

VariableDeclared commented:
Refactored and created another pull request on [container-runtime-common](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4)

[main logic is here](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4/files#diff-18083fbfdadac0ca35bd02a0aebda2c4R38)

VariableDeclared committed:
Add code path for checking no_proxy length

VariableDeclared committed:
Removed brackets

VariableDeclared committed:
Removed brackets and added config call

Corrected missing config() call.

VariableDeclared committed:
Change precedence on charm config

Changed precedence so that local charm config
wins over juju-http-proxy config.

VariableDeclared commented:
Changing to WIP as I just need to run the most recent changes through.

VariableDeclared commented:
Refactored and created another pull request on [container-runtime-common](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4)

[main logic is here](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4/files#diff-18083fbfdadac0ca35bd02a0aebda2c4R38)

Cynerva commented on reactive/docker.py:
```diff
@@ -180,7 +185,7 @@ def install():
         return False
 
-    validate_config()
+    validate_config(config())
```
I think we want to use `check_for_juju_https_proxy` here, too. As it is now, it looks like this charm will initially configure and start the docker service without including the juju-proxy vars, until a future `recycle_daemon` corrects it.

VariableDeclared replied:
Makes sense

Cynerva commented on README.md:
```diff
@@ -43,6 +43,8 @@ principal charm.
 See [config.yaml](config.yaml) for
 list of configuration options.
 
+> Note: Setting HTTP proxy values will be overriden if `juju-http-proxy` or `juju-https-proxy` settings are set on the model.
```
This is no longer true, correct?

VariableDeclared replied:
Cheers again!

Cynerva commented on README.md:
```diff
@@ -52,20 +54,24 @@ and logging.
+
+# Testing
+
+When upgrading this charm, the [`docker_runtime_version`](https://github.com/VariableDeclared/jenkins/blob/3bf2cabc7185568e6a80a137a7f083c428048683/jobs/validate-juju-https-envs.yaml#L22) of the `validate-https-proxy-envs` key needs to be updated with the new containerd version.
```
I suggest removing this from the README and removing the need to hard-code charm revs in the test.

More detailed comment: https://github.com/charmed-kubernetes/charm-containerd/pull/8#discussion_r305098886

VariableDeclared committed:
Review Changes

Added check for juju-http{s} proxy to install
method.

VariableDeclared committed:
Changed HTTP note

Missed the HTTP note. Updated this for changed
precedent.

VariableDeclared committed:
Fix grammar

Cynerva commented:
Thanks, LGTM :+1:
